src/stream/quic_lb/ngx_stream_quic_lb_retry_service.c

```diff
@@ -154,313 +154,529 @@ ngx_stream_quic_lb_gen_and_send_retry_packet(ngx_connection_t *c,
-        c, key, token, in, p - in);
+    return NGX_OK;
 }
 
-/*
- * Format: IP_ADDRESS(4B(ipv4)/16B(ipb6)) Timestamp(8B)
- * Encryption: AES-256-ECB
- */
+
 static ngx_int_t
-ngx_stream_quic_lb_new_ngx_quic_model_token(ngx_connection_t *c, ngx_str_t *token, u_char *key)
+ngx_stream_quic_lb_write_port(ngx_connection_t *c, u_char *buf,
+    ngx_int_t buf_len, size_t *out_len)
 {
-    int                   len, iv_len;
-    u_char               *data, *p, *iv;
-    ngx_msec_t            now;
-    EVP_CIPHER_CTX       *ctx;
-    const EVP_CIPHER     *cipher;
     struct sockaddr_in   *sin;
 #if (NGX_HAVE_INET6)
     struct sockaddr_in6  *sin6;
 #endif
-    u_char                in[NGX_QUIC_MAX_TOKEN_SIZE];
+    ngx_int_t             len = sizeof(uint16_t);
+
+    if (buf_len < len) {
+        ngx_log_error(NGX_LOG_ERR, c->pool->log, 0,
+                      "QUIC-LB, gen retry token error, buf len too small");
+        return NGX_ERROR;
+    }
 
     switch (c->sockaddr->sa_family) {
 
 #if (NGX_HAVE_INET6)
     case AF_INET6:
         sin6 = (struct sockaddr_in6 *) c->sockaddr;
-        len = sizeof(struct in6_addr);
-        data = sin6->sin6_addr.s6_addr;
+        buf = ngx_quic_write_uint16(buf, sin6->sin6_port);
         break;
 #endif
 
-#if (NGX_HAVE_UNIX_DOMAIN)
-    case AF_UNIX:
-        len = ngx_min(c->addr_text.len, NGX_QUIC_MAX_TOKEN_SIZE - sizeof(now));
-        data = c->addr_text.data;
+    case AF_INET:
+        sin = (struct sockaddr_in *) c->sockaddr;
+        buf = ngx_quic_write_uint16(buf, sin->sin_port);
         break;
-#endif
 
     default: /* AF_INET */
-        sin = (struct sockaddr_in *) c->sockaddr;
-        len = sizeof(in_addr_t);
-        data = (u_char *) &sin->sin_addr;
-
-        break;
+        ngx_log_error(NGX_LOG_ERR, c->pool->log, 0,
+                      "QUIC-LB, gen retry token error, only support ipv4/v6");
+        return NGX_ERROR;
     }
 
-    p = ngx_cpymem(in, data, len);
+    *out_len = len;
 
-    now = ngx_current_msec;
-    len += sizeof(now);
-    ngx_memcpy(p, &now, sizeof(now));
+    return NGX_OK;
+}
 
-    cipher = EVP_aes_256_cbc();
-    iv_len = EVP_CIPHER_iv_length(cipher);
 
-    token->len = iv_len + len + EVP_CIPHER_block_size(cipher);
-    token->data = ngx_pnalloc(c->pool, token->len);
-    if (token->data == NULL) {
+/*
+ * Defined in: https://tools.ietf.org/html/draft-ietf-quic-load-balancers-06#section-7.3
+ * Encryption: AES-128-gcm
+ * Token format:
+ * +++++++++++++++++++++++++++++++++++++++++++++++++++++
+ * | Unique Token Number(96 bit) | Key Sequence (8bit) |
+ * +++++++++++++++++++++++++++++++++++++++++++++++++++++
+ * | Token Body(...)| AEAD ICV(128bit) |
+ * +++++++++++++++++++++++++++++++++++++
+ * IV format: "Unique Token Number(96 bit)" xor "IV material(96 bit, get by key-seq and conf)"
+ * Key format: 128 bit, get by key-seq and conf
+ */
+static ngx_int_t
+ngx_stream_quic_lb_new_ngx_quic_lb_model_token(ngx_connection_t *c, ngx_str_t *token,
```

Comment:
the func name should be renamed to show this is the 
shared-retry-token's method, so as to some data struct.

Reply:
fixed

Reply:
Got that.